### PR TITLE
Add hiera defaults configuration options for all resources

### PR DESCRIPTION
**Pull request by mvisonneau**
- Added hiera defaults configuration options for all resources
- Sorted variables
- Casted them to an hash

### CODE REVIEW

**Comment by oranenj** [comment]
Should there perhaps be new tests related to this change?

Either way, looks good to me.

**Comment by wyardley** [comment]
I think this is partially covered by #1071 and similar PRs; aside from lack of tests, I think some of these resources may already have defaults that I think would get overwritten by these, e.g., https://github.com/voxpupuli/puppet-nginx/blob/master/manifests/resource/server.pp#L447-L453 (which is handled in #1071).

**Comment by oranenj** [comment]
I merged the other PR since it was smaller and had tests, so this will need a rebase for the remaining functionality.

**Comment by mvisonneau** [comment]
👋 apologies for the late update. Just rebased and added some additional checks.

**Comment by wyardley** [comment]
@mvisonneau thanks for rebasing and adding tests. It seems as if some tests are failing in CI

**Comment by bastelfreak** [comment]
Hi @mvisonneau, are you still interested in this PR? If so, please rebase against latest master. Afterwards we can check if travis is still failing.

File: manifests/init.pp [hunk 1]
```diff
@@ -151,30 +151,34 @@
...
   contain '::nginx::service'
 
-  create_resources('nginx::resource::upstream', $nginx_upstreams)
-  create_resources('nginx::resource::server', $nginx_servers, $nginx_servers_defaults)
-  create_resources('nginx::resource::location', $nginx_locations, $nginx_locations_defaults)
-  create_resources('nginx::resource::mailhost', $nginx_mailhosts, $nginx_mailhosts_defaults)
-  create_resources('nginx::resource::streamhost', $nginx_streamhosts)
-  create_resources('nginx::resource::map', $string_mappings)
-  create_resources('nginx::resource::geo', $geo_mappings)
+  create_resources( 'nginx::resource::geo', $geo_mappings, $geo_mappings_defaults )
```
Comment (bastelfreak) [inline manifests/init.pp]: Isn't this now considered bad practice? Instead we should loop over the hash. As an example: https://github.com/bastelfreak/puppet-ferm/blob/master/manifests/init.pp#L48

CC: @hunner @eputnam. What is your opinion here? We should update our review guidelines I guess.

Comment (mvisonneau) [reply manifests/init.pp]: wouldn't this change imply another request due to its nature ?

Comment (bastelfreak) [reply manifests/init.pp]: yes, let's keep it as is.

File: manifests/init.pp [hunk 2]
```diff
@@ -167,33 +167,37 @@
...
+  create_resources( 'nginx::resource::location', $nginx_locations, $nginx_locations_defaults )
+  create_resources( 'nginx::resource::mailhost', $nginx_mailhosts, $nginx_mailhosts_defaults )
+  create_resources( 'nginx::resource::map', $string_mappings, $string_mappings_defaults )
+  create_resources( 'nginx::resource::server', $nginx_servers, $nginx_servers_defaults )
+  create_resources( 'nginx::resource::streamhost', $nginx_streamhosts, $nginx_streamhosts_defaults )
+  create_resources( 'nginx::resource::upstream', $nginx_upstreams, $nginx_upstreams_defaults )
```
Comment (alexjfisher) [inline manifests/init.pp]: As per https://github.com/voxpupuli/puppet-nginx/pull/1309, good luck creating `nginx::resource::upstream` resources with defaults of type `Nginx::UpstreamMemberDefaults`...

**Pull request merged**